addons/edi_stock/models/edi_move_cancellation_report_document.py

```diff
@@ -0,0 +1,195 @@
+"""EDI stock move request records"""
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class IrModuleModule(models.Model):
+    _inherit = "ir.module.module"
+
+    def _is_sale_stock_installed(self):
+        """
+        Returns bool() sale_stock module existing and being in state 'installed'.
+        With more time this be separated out into a bridge module
+        """
+        sale_stock_module = self.search([("name", "=", "sale_stock")])
+        return bool(sale_stock_module and sale_stock_module.state == "installed")
+
+
+class StockMove(models.Model):
+    """Extend ``stock.move`` to include the EDI Stock Cancellation document"""
+
+    _inherit = "stock.move"
+
+    edi_stock_cancellation_doc_id = fields.Many2one(
+        "edi.document",
+        string="EDI Stock Cancellation Document",
+        required=False,
+        readonly=True,
+        index=True,
+    )
+
+
+class EdiDocument(models.Model):
+    """Extend ``edi.document`` to include stock move cancellation records"""
+
+    _inherit = "edi.document"
+
+    move_cancellation_ids = fields.One2many(
+        "edi.record.stock.cancellation",
+        "doc_id",
+        string="Move Cancellation Records",
+    )
+
+
+class EdiStockCancellationDocument(models.AbstractModel):
+    """Stock Cancellation document model
+
+    Stock Cancellation documents contain sets of moves
+    (grouped by sale/pick partner & picking_type) which have been cancelled
+    since the previous EDI export.
+    """
+
+    _name = "edi.document.stock.cancellation"
+    _inherit = "edi.move.report.document"
+    _description = "Stock Cancellations"
+
+    _edi_move_report_via = "edi_stock_cancellation_doc_id"
+
+    @api.model
+    def move_report_domain(self, _doc):
+        """
+        Override to only include cancelled moves for picking types which have
+        u_enable_pick_cancellation_email_notifs enabled.
+        _edi_move_report_via functionality is kept to ensure moves are not reported on twice
+        """
+        StockPickingType = self.env["stock.picking.type"]
+        picking_types_to_consider = StockPickingType.search(
+            [("u_enable_pick_cancellation_email_notifs", "=", True)]
+        )
+        domain = [
+            ("state", "=", "cancel"),
+            (self._edi_move_report_via, "=", False),
+            ("picking_id.picking_type_id", "in", picking_types_to_consider.ids),
+        ]
+        return domain
+
+    @api.model
+    def move_report_list(self, doc, moves):
+        """
+        Inherit to group moves by sale & picking_type to result
+        in a cancellation record for each unique combination
+        """
+        IrModuleModule = self.env["ir.module.module"]
+        res = super().move_report_list(doc, moves)
+        grouped_moves = []
+
+        if IrModuleModule._is_sale_stock_installed():
```

Comment:
It would be better to call another function to handle all of this. Preferably only handling the case without sale orders. Then create a new module edi_sale_stock where we override such function to handle the sale order case.